core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreAggregateCaseAggregations.java

```diff
@@ -280,19 +288,66 @@ private ProjectNode createPreProjection(
         return new ProjectNode(context.getIdAllocator().getNextId(), source, assignments.build());
     }
 
-    private Map<PreAggregationKey, PreAggregation> getPreAggregations(List<CaseAggregation> aggregations, Context context)
```

Comment:
instead of removing `getPreAggregations`, would it be possible to construct and return `Map<PreAggregationKey, PreAggregation>` within `getPreAggregations`, while `getPreAggregations` would:
1. internally group `CaseAggregations` by `PreAggregationKey` into multimap
2. build `Map<PreAggregationKey, PreAggregation>` with union `IF` expressions

I think then the reminder of `apply` wouldn't need to change

Reply:
We are indeed doing the internal grouping within `createPreAggregations`. As explained in other comment, however, we cannot use `PreAggregationKey` for grouping because the result sections are all different. Think of Q66 as an example:
```
"sum"((CASE WHEN ("d_moy" = 1) THEN IF("d_moy" = 1, "cs_sales_price" * "cs_quantity") ELSE 0 END)) "jan_sales",
"sum"((CASE WHEN ("d_moy" = 2) THEN IF("d_moy" = 2, "cs_sales_price" * "cs_quantity") ELSE 0 END)) "feb_sales"
```

Since after `createPreAggregations`, the keys of `preAggregations` possibly contain merged projections, they cannot be used for rebuilding the downstream projection as the lookup would fail to find the matching key.

Reply:
I don't understand this example fully:
```
"sum"((CASE WHEN ("d_moy" = 1) THEN IF("d_moy" = 1, "cs_sales_price" * "cs_quantity") ELSE 0 END)) "jan_sales",
"sum"((CASE WHEN ("d_moy" = 2) THEN IF("d_moy" = 2, "cs_sales_price" * "cs_quantity") ELSE 0 END)) "feb_sales"
```

pre-projection in this case should be `IF("d_moy" = 1 OR "d_moy=2", "cs_sales_price" * "cs_quantity")`

`PreAggregationKey` would be `function := sum, projection := "cs_sales_price" * "cs_quantity"`, so the translation would be:
```
sum(case when(d_moy = 1) then pre_aggregation)
  sum(pre_projection, group by=d_moy)
    pre_projection := IF("d_moy" = 1 OR "d_moy=2", "cs_sales_price" * "cs_quantity")
```

What your code would produce for the case above?


What about:
```
"sum"((CASE WHEN ("d_moy" = 1) THEN IF("d_moy" = 1, "cs_sales_price" * "cs_quantity") ELSE 0 END)) "jan_sales",
"max"((CASE WHEN ("d_moy" = 2) THEN IF("d_moy" = 2, "cs_sales_price" * "cs_quantity") ELSE 0 END)) "feb_sales"
```
?

Reply:
The thing is this input
```
"sum"((CASE WHEN ("d_moy" = 1) THEN "cs_sales_price" * "cs_quantity" ELSE 0 END)) "jan_sales",
"sum"((CASE WHEN ("d_moy" = 2) THEN "cs_sales_price" * "cs_quantity" ELSE 0 END)) "feb_sales"
```
already becomes
```
"sum"((CASE WHEN ("d_moy" = 1) THEN IF("d_moy" = 1, "cs_sales_price" * "cs_quantity") ELSE 0 END)) "jan_sales",
"sum"((CASE WHEN ("d_moy" = 2) THEN IF("d_moy" = 2, "cs_sales_price" * "cs_quantity") ELSE 0 END)) "feb_sales"
```
after we perform "extractCaseAggregations" initially.

So the following logic like `createPreAggregations` needs to work with the "IF'ed" projections in the `PreAggregationKey`. That's what caused the complexity.